lighthouse-core/lib/dependency-graph/base-node.js

```diff
@@ -168,96 +168,79 @@ class BaseNode {
+        node => node._dependencies.filter(parent => !idsToIncludedClones.has(parent.id))
+      );
     });
 
+    // Copy dependencies between nodes.
     rootNode.traverse(originalNode => {
-      if (!shouldIncludeNode(originalNode)) return;
-      const clonedNode = idToNodeMap.get(originalNode.id);
+      const clonedNode = idsToIncludedClones.get(originalNode.id);
+      if (!clonedNode) return;
 
       for (const dependency of originalNode._dependencies) {
-        const clonedDependency = idToNodeMap.get(dependency.id);
+        const clonedDependency = idsToIncludedClones.get(dependency.id);
+        if (!clonedDependency) throw new Error('Dependency somehow not cloned');
         clonedNode.addDependency(clonedDependency);
       }
     });
 
-    if (!idToNodeMap.has(this.id)) throw new Error('Cloned graph missing node');
-    return idToNodeMap.get(this.id);
+    const clonedThisNode = idsToIncludedClones.get(this.id);
+    if (!clonedThisNode) throw new Error('Cloned graph missing node');
+    return clonedThisNode;
   }
 
   /**
-   * Traverses all paths in the graph, calling iterator on each node visited. Decides which nodes to
-   * visit with the getNext function.
-   * @param {function(Node, Node[]): void} iterator
-   * @param {function(Node): Node[]} getNext
+   * Traverses all connected nodes in BFS order, calling `callback` exactly once
+   * on each. `traversalPath` is the shortest (though not necessarily unique)
+   * path from `node` to the root of the iteration.
+   *
+   * The `getNextNodes` function takes a visited node and returns which nodes to
+   * visit next. It defaults to returning the node's dependents.
+   * @param {(node: Node, traversalPath: Node[]) => void} callback
+   * @param {function(Node): Node[]} [getNextNodes]
    */
-  _traversePaths(iterator, getNext) {
-    const stack = [[/** @type {Node} */(/** @type {BaseNode} */(this))]];
-    while (stack.length) {
-      /** @type {Node[]} */
-      // @ts-ignore - stack has length so it's guaranteed to have an item
-      const path = stack.shift();
-      const node = path[0];
-      iterator(node, path);
-
-      const nodesToAdd = getNext(node);
-      for (const nextNode of nodesToAdd) {
-        stack.push([nextNode].concat(path));
-      }
+  traverse(callback, getNextNodes) {
+    if (!getNextNodes) {
+      getNextNodes = node => node.getDependents();
     }
-  }
 
-  /**
-   * Traverses all connected nodes exactly once, calling iterator on each. Decides which nodes to
-   * visit with the getNext function.
-   * @param {function(Node, Node[]): void} iterator
-   * @param {function(Node): Node[]} [getNext] Defaults to returning the dependents.
-   */
-  traverse(iterator, getNext) {
-    if (!getNext) {
-      getNext = node => node.getDependents();
-    }
+    const visited = new Set([this.id]);
+    const queue = [[/** @type {Node} */(/** @type {BaseNode} */(this))]];
```

Comment:
do we still need this? I'm surprised we need to cast `BaseNode` for `this` hopefully this is something that was fixed :)

Reply:
> do we still need this? 

Annoyingly, yes. `this` *extends* `BaseNode`, but the lie is that it's a `Node`. Humans know the only types of `BaseNode` are `CPUNode | NetworkNode` so the cast to `Node` is ok, but the compiler accounts for the fact that we could be calling on a different class derived from `BaseNode`, so it's not necessarily compatible with `Node`.

So to cast while following the type assertion rules, we have to upcast to `BaseNode` and then downcast to `Node`. The other options is an assertion and a `@ts-ignore`...maybe that's more honest (albeit also ugly).

Reply:
> Humans know the only types of `BaseNode` are `CPUNode | NetworkNode` so

I mean, even that's not entirely true. In tests we call `traverse()` with a graph of `BaseNode` nodes, but in real code it's much more convenient to only get `CPUNode | NetworkNode` back, so we should continue to pretend it's true :)